apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+    * @param _finishedEvidence Whether the challenger is finished submitting evidence with the challenge context
+    * @param _context Link to a human-readable text providing context for the challenge
+    */
+    function challengeAction(uint256 _actionId, uint256 _settlementOffer, bool _finishedEvidence, bytes _context) external {
+        Action storage action = _getAction(_actionId);
+        require(_canChallenge(_actionId, action), ERROR_CANNOT_CHALLENGE_ACTION);
+
+        (IDisputable disputable, CollateralRequirement storage requirement) = _getDisputableFor(action);
+        require(_canPerformChallenge(disputable, msg.sender), ERROR_SENDER_CANNOT_CHALLENGE_ACTION);
+        require(_settlementOffer <= requirement.actionAmount, ERROR_INVALID_SETTLEMENT_OFFER);
+
+        // TODO: implement try catch
+        uint256 challengeId = _createChallenge(_actionId, action, msg.sender, requirement, _settlementOffer, _finishedEvidence, _context);
+        action.currentChallengeId = challengeId;
+        disputable.onDisputableActionChallenged(action.disputableActionId, challengeId, msg.sender);
+        emit ActionChallenged(_actionId, challengeId);
+    }
+
+    /**
+    * @notice Settle challenged action #`_actionId`, accepting the settlement offer
+    * @param _actionId Identification number of the action to be settled
+    */
+    function settle(uint256 _actionId) external {
+        (Action storage action, Challenge storage challenge, uint256 challengeId) = _getChallengedAction(_actionId);
+        address submitter = action.submitter;
+
+        if (msg.sender == submitter) {
+            require(_canSettle(_actionId, challenge), ERROR_CANNOT_SETTLE_ACTION);
+        } else {
+            require(_canClaimSettlement(_actionId, challenge), ERROR_CANNOT_SETTLE_ACTION);
+        }
+
+        (IDisputable disputable, CollateralRequirement storage requirement) = _getDisputableFor(action);
+        uint256 actionCollateral = requirement.actionAmount;
+        uint256 settlementOffer = challenge.settlementOffer;
+
+        // The settlement offer was already checked to be up-to the collateral amount upon challenge creation
+        // However, we cap it to collateral amount to be safe
+        // With this, we can avoid using SafeMath to calculate `unlockedAmount`
+        uint256 slashedAmount = settlementOffer >= actionCollateral ? actionCollateral : settlementOffer;
+        uint256 unlockedAmount = actionCollateral - slashedAmount;
+
+        address challenger = challenge.challenger;
+        _unlockAndSlashBalance(requirement.staking, submitter, unlockedAmount, challenger, slashedAmount);
+        _transfer(requirement.token, challenger, requirement.challengeAmount);
+        _transfer(challenge.arbitratorFeeToken, challenger, challenge.arbitratorFeeAmount);
+
+        challenge.state = ChallengeState.Settled;
+        disputable.onDisputableActionRejected(action.disputableActionId);
+        emit ActionSettled(_actionId, challengeId);
```

Comment:
⚠️ Want to be careful here:

- Is `onDisputableActionRejected` the correct event? It's a bit confusing.
- Should we also emit an event that the action was rejected?
- Should we reset the currentChallengeId back to 0?

Reply:
> Is `onDisputableActionRejected` the correct event? It's a bit confusing.

I lifted this as another assumption, I mean treating settled actions as rejected ones. 

> Should we also emit an event that the action was rejected?

I think it will make it ambiguous if we do that. I know we are deciding to sum it up for the disputable app treating this case as rejected, but it's simply because we will be forcing disputables to handle another scenario that in 99.9% of the cases will be handled in the same way as rejected actions.

> Should we reset the `currentChallengeId` back to 0?

Already discussed [here](https://github.com/aragon/aragon-apps/pull/1133#discussion_r439650486)

Reply:
> I lifted this as another assumption, I mean treating settled actions as rejected ones.

Sounds good; I think this is a reasonable assumption.

My confusion from before was actually between the "challenge accepted" vs. "action accepted" terminology—I think this has been cleared up now!

> Should we also emit an event that the action was rejected?

Agreed that we can leave it out. We can treat the `ActionAccepted`, `ActionVoided`, and `ActionRejected` events as dependent on a dispute being created.